samples/csharp/skill/SkillSample.FunctionalTests/Bot/TestBotClient.cs

```
@@ -0,0 +1,242 @@
+            // why we add this custom code.
+            using (HttpClient client = new HttpClient())
+            {
+                HttpRequestMessage request = new HttpRequestMessage(HttpMethod.Post, $"https://directline.botframework.com/v3/directline/tokens/generate");
+                request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", config.DirectLineSecret);
+                request.Content = new StringContent(
+                    JsonConvert.SerializeObject(new
+                    {
+                        User = new { Id = this.user },
+                        TrustedOrigins = new string[]
+                        {
+                            OriginHeaderValue
+                        }
+                    }), Encoding.UTF8, "application/json");
+
+                using (var response = client.SendAsync(request).GetAwaiter().GetResult())
+                {
+                    if (response.IsSuccessStatusCode)
+                    {
+                        // Extract token from response
+                        var body = response.Content.ReadAsStringAsync().GetAwaiter().GetResult();
+                        this.token = JsonConvert.DeserializeObject<DirectLineToken>(body).Token;
+                        this.conversationId = JsonConvert.DeserializeObject<DirectLineToken>(body).ConversationId;
+
+                        // Create directline client from token
+                        this.directLineClient = new DirectLineClient(token);
+
+                        // From now on, we'll add an Origin header in directline calls, with
+                        // the trusted origin we sent when acquiring the token as value.
+                        directLineClient.HttpClient.DefaultRequestHeaders.Add(OriginHeaderKey, OriginHeaderValue);
+                    }
+                    else
+                    {
+                        throw new Exception("Failed to acquire directline token");
```

It would be better to return an exception that indicates the response returned so that error information isn't lost.